shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/RouteCardData.kt

```diff
@@ -306,90 +306,51 @@ data class RouteCardData(
-
-            val parentToAllStopIds = mutableMapOf<String, MutableSet<String>>()
-
-            globalData.run {
-                stopIds.forEach { stopId ->
-                    val stop = stops[stopId] ?: return@forEach
-                    val patternsByRouteOrLine =
-                        patternsByRouteOrLine(stopId, globalData)
-                            // filter out a route if we've already seen all of its patterns
-                            .filterNot {
-                                routePatternsUsed.containsAll(
-                                    it.value.map { pattern -> pattern.id }.toSet()
-                                )
-                            }
-                    routePatternsUsed.addAll(
-                        patternsByRouteOrLine.flatMap { it.value }.map { it.id }
-                    )
-
-                    stop.parentStationId?.let { parentStationId ->
-                        parentToAllStopIds
-                            .getOrPut(parentStationId) { mutableSetOf(parentStationId) }
-                            .add(stop.id)
+            // A map of a stop to itself and any children.
+            // for standalone stops, an entry will be <standaloneStop, [standaloneStopId]>.
+            // for stations, an entry will be <station, [stationId, child1Id, child2Id, etc.]>
+            val parentToAllStops: Map<Stop, Set<String>> =
+                stopIds
+                    .mapNotNull { stopId ->
+                        val stop = globalData.stops[stopId]
+                        if (stop != null) {
+                            Pair(stop.resolveParent(globalData), stop)
+                        } else {
+                            null
+                        }
                     }
+                    .groupBy({ it.first }, { it.second.id })
+                    .mapValues { it.value.toSet().plus(it.key.id) }
 
-                    val stopOrParent = stop.resolveParent(stops)
-
-                    for ((routeOrLine, routePatterns) in patternsByRouteOrLine) {
-                        val routeStops = patternsGrouped.getOrPut(routeOrLine) { mutableMapOf() }
-                        val patternsForStop = routeStops.getOrPut(stopOrParent) { mutableListOf() }
-                        patternsForStop += routePatterns
-                    }
-                }
-            }
+            val patternsGrouped = patternsGroupedByLineOrRouteAndStop(globalData, parentToAllStops)
 
             val builderData =
                 patternsGrouped
-                    .map { byLineOrRoute ->
-                        val key =
-                            when (val cardType = byLineOrRoute.key) {
-                                is LineOrRoute.Line -> cardType.line.id
-                                is LineOrRoute.Route -> cardType.route.id
-                            }
-                        // TODO: Directions list with actual values.
+                    .map { (lineOrRoute, patternsByStop) ->
```

Comment:
I think I forgot on my initial writing of this that you can de-structure map entries like this!